compiler/noirc_evaluator/src/ssa/opt/inlining.rs

```diff
@@ -392,14 +479,24 @@ impl<'function> PerFunctionContext<'function> {
                 Instruction::Call { func, arguments } => match self.get_function(*func) {
                     Some(func_id) => {
                         let function = &ssa.functions[&func_id];
-                        // If we have not already finished the flattening pass, functions marked
-                        // to not have predicates should be marked as entry points unless we are inlining into brillig.
-                        let entry_point = &ssa.functions[&self.context.entry_point];
-                        let no_predicates_is_entry_point =
-                            self.context.no_predicates_is_entry_point
-                                && function.is_no_predicates()
-                                && !matches!(entry_point.runtime(), RuntimeType::Brillig);
-                        if function.runtime().is_entry_point() || no_predicates_is_entry_point {
+
+                        let should_retain_call =
+                            if let RuntimeType::Acir(inline_type) = function.runtime() {
+                                // If the called function is acir, we inline if it's not an entry point
+
+                                // If we have not already finished the flattening pass, functions marked
+                                // to not have predicates should be marked as entry points.
+                                let no_predicates_is_entry_point =
+                                    self.context.no_predicates_is_entry_point
+                                        && function.is_no_predicates();
+                                inline_type.is_entry_point() || no_predicates_is_entry_point
+                            } else {
+                                // If the called function is brillig, we inline only if it's into brillig and the function is not recursive
+                                ssa.functions[&self.context.entry_point].runtime()
+                                    != RuntimeType::Brillig
+                                    || self.context.recursive_functions.contains(&func_id)
+                            };
```

Comment:
Can you break this out into a helper function called `should_inline_call` or similar? I think it's getting large enough now that it'd simplify reading `inline_block_instructions` if it were cut out